src/main/scala/com/horizon/exchangeapi/AuthenticationSupport.scala

```diff
@@ -1012,7 +1112,36 @@ trait AuthenticationSupport extends ScalatraBase {
     override def isPublic: Boolean = if (all) return true else return AuthCache.patterns.getIsPublic(id).getOrElse(false)
   }
   case class TAction(id: String = "") extends Target    // for post rest api methods that do not target any specific resource (e.g. admin operations)
+}
+
+/** Authenticates the client credentials and then checks the ACLs for authorization. */
+trait AuthenticationSupport extends ScalatraBase with AuthSupport {
+  // We could add a before action with befor() {}, but sometimes they need to pass in user/pw, and sometimes id/token
+  // I tried using code from http://www.scalatra.org/2.4/guides/http/authentication.html, but it throws an exception.
+
+  def db: Database      // get access to the db object in ExchangeApiApp
+  implicit def logger: Logger    // get access to the logger object in ExchangeApiApp
 
+  var migratingDb = false     // used to lock everyone out during db migration
+  def isDbMigration = migratingDb
+  // def setDbMigration(dbMigration: Boolean): Unit = { migratingDb = dbMigration }
+
+  /* This is the main function that will be used in the routes, i.e., it
+   * replaces the usage of credsAndLog().authenticate(). It follows the
+   * previous behavior of returning an authenticated Identity, which can
+   * be used for authorization, or halting the request due to invalid
+   * credentials.
+   */
+  def authenticate(anonymousOk: Boolean = false, hint: String = ""): Identity = {
+    val loginCtx = new LoginContext(
+      "ExchangeApiLogin",
+      new ExchCallbackHandler(RequestInfo(request, params, isDbMigration, anonymousOk, hint))
+    )
+    for (failure <- Try(loginCtx.login()).failed) {
+      halt(HttpCode.BADCREDS, ApiResponse(ApiResponseType.BADCREDS, "invalid credentials"))
+    }
+    loginCtx.getSubject.getPrivateCredentials(classOf[Identity]).asScala.head
+  }
```

Comment:
For JAAS, the `LoginContext` is what you use to attempt to login a user and get a `Subject` back. It takes care of creating the `LoginModule`s and calling them. It can be configured a few ways, the simplest is to have a config file that specifies which LoginModules to use, which is what I've been doing.